### PR TITLE
Add support for button entities

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

Implement `button` from Home Assistant

As a simple example, I have implemented `restart` and `template` buttons. Example YAML below.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1681

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
button:
  - platform: restart
  	name: Restart
  - platform: template
  	name: Template button
    on_press:
      then:
        - logger.log: "Button pressed"

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by dudanov** [comment]
@jesserockz Jesse, thanks for the new component! How can I instantiate the ``Button`` class when it is abstract? It would be nice to be able to instantiate and use ``Button`` for simple automation tasks.

**Comment by jesserockz** [comment]
> @jesserockz Jesse, thanks for the new component! How can I instantiate the `Button` class when it is abstract? It would be nice to be able to instantiate and use `Button` for simple automation tasks.

I will also make a template button for these cases. 
So it would be like:
```yaml
button: 
  - platform: template
    name: blah button
    on_press:
      then: 
        - logger.log: Pressed Button
```

File: esphome/components/api/api_connection.h [hunk 1]
```diff
@@ -50,6 +50,10 @@ class APIConnection : public APIServerConnection {
   bool send_switch_info(switch_::Switch *a_switch);
   void switch_command(const SwitchCommandRequest &msg) override;
 #endif
+#ifdef USE_BUTTON
+  bool send_button_info(button::Button *button);
+  void button_command(const ButtonCommandRequest &msg) override;
+#endif
```
Comment (oxan) [inline esphome/components/api/api_connection.h]: This file uses (mostly) the same order as the proto file, so can we move this down to be consistent?

File: esphome/components/restart/button/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+restart_ns = cg.esphome_ns.namespace("restart")
+RestartButton = restart_ns.class_("RestartButton", button.Button, cg.Component)
+
+CONFIG_SCHEMA = button.BUTTON_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(RestartButton),
+        cv.Optional(CONF_ICON, default=ICON_RESTART): button.icon,
+        cv.Optional(
+            CONF_ENTITY_CATEGORY, default=ENTITY_CATEGORY_CONFIG
+        ): cv.entity_category,
```
Comment (oxan) [inline esphome/components/restart/button/__init__.py]: Maybe it's better to go to a sensor-style schema-generating-function that takes the defaults as arguments, so this validation doesn't have to be duplicated for every button.

**Review by oxan — APPROVED** [review]
LGTM in general, two minor nits:

File: esphome/components/api/api_connection.h (hunk 1 above)
Comment (jesserockz) [reply esphome/components/api/api_connection.h]: Done

File: esphome/components/restart/button/__init__.py (hunk 2 above)
Comment (jesserockz) [reply esphome/components/restart/button/__init__.py]: Done

**Review by oxan — APPROVED** [review]
Didn't test it, but code looks good.

**Pull request merged**